### ISSUE TITLE
iMac Pro(2017) & MacBook Pro(2019) using plaidML slow as CPU

### ISSUE THREAD

**Issue by AEM90** [plaidml/plaidml#903]
Hi Dear Community,
im new to this community and i'm bringing an interesting bug.

I am doing my master thesis on Machine Learning and i want to predict one value out of many. Since my data is time series data i use a LSTM. I installed the PlaidML packages as explained in the tutorial below.
https://towardsdatascience.com/gpu-accelerated-machine-learning-on-macos-48d53ef1b545

After installing everything i tried out the example which is appended. This one works without any 
_INFO:plaidml:Analyzing Ops: 87908 of 91336 operations complete_
step, but it does need more time to process than the tutorial users needs. Which is already a problem since my imac and my macbook have both better specs than he has.

When now i use my LSTM with plaidML and Keras it does use the GPU, it does tell me its using PlaidML, but is super slow. What am i doing wrong?

I attached my code with comments. It would be lovely if you could provide me some help.

Thanks!

CODE: 
import plaidml.keras

plaidml.keras.install_backend()

import os
# os.environ["KERAS_BACKEND"] = "plaidml.keras.backend"

import keras
keras.backend.backend()

from keras.models import Sequential
from keras.layers import LSTM
from keras.layers import Dense
from keras import optimizers
from sklearn.model_selection import train_test_split
import glob, re, math, decimal, os, copy
from sklearn import preprocessing

import methoden
import numpy as np
import pandas as pd
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "last"

# Import Data
# List of Dataframes which include the signals for my LSTM
preuss = list()
for i in range(0,9):
  preuss.append(pd.read_csv('preuss' + str(i) + '.csv', sep='\t'))

#prepare data: dataframe to 3d numpy array
tmp_X = []
tmp_y = []
for i in range(len(preuss)):
    # split into sequences
    n_steps = 300
    X3d, y3d = methoden.split_sequences(preuss[i], n_steps)
    tmp_X.append(X3d)
    tmp_y.append(y3d)
X = np.concatenate(tmp_X)
y = np.concatenate(tmp_y)

# define input sequence, create training batches
n_steps = X.shape[1]
n_features = X.shape[2]
print(X.shape, 'samples, timesteps, features')
# split in train and test data
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size=0.2, random_state=42)
print('total amount of training samples:', len(train_X))
print('total amount of test samples:', len(test_X))

#normalize data
scaler_y = preprocessing.MinMaxScaler(feature_range=(0, 1))
scaler_x = preprocessing.MinMaxScaler()
for i in range(len(train_X)):
    train_X[i] = scaler_x.fit_transform(train_X[i])
for i in range(len(test_X)):
    test_X[i] = scaler_x.transform(test_X[i])

train_y, scaleVal_train_y = methoden.normalize_signals(train_y)

# define model: im not very sure if im doing it well. INPUT 7 Features. OUTPUT 1 Feature
model = Sequential()
model.add(LSTM(100, activation='tanh', return_sequences=True, input_shape=(X.shape[1], X.shape[2])))
model.add(LSTM(100, activation='sigmoid', return_sequences=True))
# model.add(LSTM(256, activation='relu', return_sequences=True))
model.add(LSTM(100, activation='tanh'))
model.add(Dense(units=1, input_shape=(1,), activation='relu'))

sgd = optimizers.SGD(lr=0.01, decay=1e-4, momentum=0.7, nesterov=True)
model.compile(optimizer='sgd', loss='mse')

history = model.fit(train_X, train_y, epochs=200, verbose=1)

model_result = model.predict(test_X, verbose=0)
model_result_clean = scaler_x.inverse_transform(model_result)
print(abs(test_y - model_result_clean))

**Comment by krinkere** [comment]
same here... just tried it on mine and it is slow

**Comment by brianretford** [comment]
LSTMs are slow in plaidml v0.x in training, due to excessive memory consumption and sub-optimal embedding support. We will fix this in v1 - @krinkere for your 2012 macbook I'm not sure how much acceleration to expect. In general wins are usually seen with fairly dense networks and reasonable batch sizes (CNNs on AMDs)

**Comment by AEM90** [comment]
@brianretford The v1 you mentioned is this one: https://github.com/plaidml/plaidml ? (plaidml-v1)

**Comment by joelfirenze** [comment]
Having the same issue with similar code for a iMac Retina 5K, 27-inch, late 2014
AMD Radeon R9 M295X 4GB
4GHz Intel Core i7
using Plaid 0.7, running a LSTM
Keras 2.2.4
non Plaidml - 1 epoch, ~11mins
Plaidml - 1 epoch ~ 108mins
